File: app/routers/factor_screening.py
"""因子打分选股（子项目 2a）Web 接口（异步任务）。

任务提交后立即返回 task_id，真正的选股计算通过 FastAPI `BackgroundTasks`
在后台执行；任务状态记录在 MongoDB `factor_screen_tasks` 集合中（见
`app/services/factor_screening_service.set_task_status`/`get_task_status`）。
属主校验/BackgroundTasks 写法照 `app/routers/backtest.py`。
"""
import uuid
from typing import Any, Dict

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException

from app.routers.auth_db import get_current_user
from app.services import factor_screening_service as svc
from tradingagents.factor import FACTORS

router = APIRouter(prefix="/api/factor-screen", tags=["factor-screen"])


@router.get("/factors")
async def list_factors(user: dict = Depends(get_current_user)):
    """返回可选因子元信息（无需鉴权数据敏感性低，但仍加 get_current_user 保持一致）。"""
    data = [
        {"key": k, "name": m["name"], "category": m["category"],
         "default_direction": m["default_direction"]}
        for k, m in FACTORS.items()
    ]
    return {"success": True, "data": data}


def _validate(payload: Dict[str, Any]) -> None:
    """选股请求参数校验，非法参数直接 400，不进入后台任务。"""
    factors = payload.get("factors") or []
    if not factors:
        raise HTTPException(status_code=400, detail="至少选择一个因子")
    for f in factors:
        if f.get("key") not in FACTORS:
            raise HTTPException(status_code=400, detail=f"未知因子: {f.get('key')}")
        if not (f.get("weight", 0) > 0):
            raise HTTPException(status_code=400, detail="因子权重必须为正")
    if payload.get("top_n", 0) <= 0:
        raise HTTPException(status_code=400, detail="top_n 必须为正")


@router.post("/run")
async def run(
    payload: Dict[str, Any],
    background_tasks: BackgroundTasks,
    user: dict = Depends(get_current_user),
):
    """提交一次因子选股任务，返回 task_id，实际计算在后台执行。"""
    _validate(payload)

    user_id = user["id"]
    task_id = uuid.uuid4().hex

    await svc.ensure_db()
    await svc.set_task_status(task_id, "running", user_id=user_id)

    async def _run_task():
        try:
            await svc.ensure_db()
            await svc.run_screen_task(task_id, user_id, payload)
            await svc.set_task_status(task_id, "done", user_id=user_id)
        except Exception as e:  # noqa: BLE001
            await svc.set_task_status(task_id, "failed", error=str(e), user_id=user_id)

    background_tasks.add_task(_run_task)
    return {"success": True, "data": {"task_id": task_id}}


@router.get("/status/{task_id}")
async def status(task_id: str, user: dict = Depends(get_current_user)):
    """查询选股任务状态。

    属主校验：任务不存在 或 属于其他用户，统一返回 404（不用 403），
    避免向非本人泄露“该 task_id 确实存在”这一信息。
    """
    await svc.ensure_db()
    task_status = await svc.get_task_status(task_id)
    if not task_status or task_status.get("user_id") != user["id"]:
        raise HTTPException(status_code=404, detail="任务不存在")
    return {"success": True, "data": task_status}


@router.get("/result/{task_id}")
async def result(task_id: str, user: dict = Depends(get_current_user)):
    """查询选股结果，任务未完成/不存在/属于其他用户时统一返回 404。"""
    await svc.ensure_db()
    res = await svc.get_result(task_id)
    if not res or res.get("user_id") != user["id"]:
        raise HTTPException(status_code=404, detail="结果不存在")
    return {"success": True, "data": res}


@router.get("/history")
async def history(
    limit: int = 20,
    skip: int = 0,
    user: dict = Depends(get_current_user),
):
    """查询当前用户的历史选股记录。"""
    await svc.ensure_db()
    data = await svc.get_history(user["id"], limit, skip)
    return {"success": True, "data": data}

File: frontend/src/api/factorScreening.ts

/**
 * 多因子选股API
 */

import { request, type ApiResponse } from './request'

// 单个因子的打分配置：key 为因子标识，weight 为正数权重，direction 为打分方向
export interface FactorConfigItem {
  key: string
  weight: number
  direction: 'asc' | 'desc'
}

// 选股域筛选条件，均为可选：不传/为空表示该维度不过滤
export interface FactorUniverse {
  exclude_st?: boolean
  exclude_new?: boolean
  industries?: string[]
  mv_min?: number
  mv_max?: number
}

// 因子选股请求参数（与后端 app/routers/factor_screening.py `_validate` 校验的字段一致）
export interface FactorScreenRunRequest {
  factors: FactorConfigItem[]
  universe: FactorUniverse
  top_n: number
}

// 因子选股任务创建响应
export interface FactorScreenTaskResponse {
  task_id: string
  status?: string
  message?: string
  [key: string]: any
}

// 因子选股任务状态
export interface FactorScreenStatus {
  task_id: string
  status: 'running' | 'done' | 'failed'
  progress?: number
  message?: string
  error?: string
  [key: string]: any
}

// 因子选股历史查询参数
export interface FactorScreenHistoryParams {
  page?: number
  page_size?: number
  status?: string
  [key: string]: any
}

// 多因子选股API
export const factorApi = {
  // 获取可用因子列表
  factors(): Promise<ApiResponse<any>> {
    return request.get('/api/factor-screen/factors')
  },

  // 发起选股任务
  run(payload: FactorScreenRunRequest): Promise<ApiResponse<FactorScreenTaskResponse>> {
    return request.post('/api/factor-screen/run', payload)
  },

  // 查询选股任务状态
  status(taskId: string): Promise<ApiResponse<FactorScreenStatus>> {
    return request.get(`/api/factor-screen/status/${taskId}`)
  },

  // 获取选股结果
  result(taskId: string): Promise<ApiResponse<any>> {
    return request.get(`/api/factor-screen/result/${taskId}`)
  },

  // 获取选股历史
  history(params?: FactorScreenHistoryParams): Promise<ApiResponse<any>> {
    return request.get('/api/factor-screen/history', { params })
  }
}

export default factorApi
